<?xml version="1.0"?>
<!--
Software License Agreement (BSD)

\file      stereo.launch
\authors   Michael Hosmar <mhosmar@clearpathrobotics.com>
\copyright Copyright (c) 2018, Clearpath Robotics, Inc., All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that
the following conditions are met:
 * Redistributions of source code must retain the above copyright notice, this list of conditions and the
   following disclaimer.
 * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the 
   following disclaimer in the documentation and/or other materials provided with the distribution.
 * Neither the name of Clearpath Robotics nor the names of its contributors may be used to endorse or promote
   products derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WAR-
RANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, IN-
DIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-->
<launch>


  <!-- Common parameters -->
  <arg name="camera_name" default="stereorig" />

  <!-- When unspecified, the driver will use the default framerate as given by the
      camera itself. Use the parameter 'control_frame_rate' to enable manual frame 
      rate control, and 'frame_rate' to set the frame rate value. -->
  <arg name="frame_rate" default="20" />

  <arg name="left_camera_serial" default="21387972" />
  <arg name="left_camera_calibrated" default="0" />

  <arg name="right_camera_serial" default="21502646" />
  <arg name="right_camera_calibrated" default="0" />

  <arg name="run_stereo_image_proc" default="false" />

  <arg name="encoding" default="RGB8Packed" />
  <arg name="color_balance" default="Continuous" />
  <!-- Off, Once, or Continuous -->

  <!-- Nodelet manager -->
  <!-- Both cameras are not loaded into one nodelet manager to avoid the namespacing issue. -->
  <node pkg="nodelet" type="nodelet" name="camera_nodelet_manager" args="manager" output="screen"/>

  <group ns="$(arg camera_name)">
    <group ns="left">
      <!-- Camera nodelet -->
      <node pkg="nodelet" type="nodelet" name="camera_nodelet" args="load spinnaker_camera_driver/SpinnakerCameraNodelet /camera_nodelet_manager">

        <param name="mode" value="primary" />
        <param name="frame_id" value="camera_left" />
        <param name="serial" value="$(arg left_camera_serial)" />

        <!-- Frame rate -->
        <param name="acquisition_frame_rate_enable" value="true" />
        <param name="acquisition_frame_rate" value="$(arg frame_rate)" />

        <param name="image_format_color_coding" value="$(arg encoding)" />
        <param name="auto_white_balance" value="$(arg color_balance)" />

        <!-- Use the camera_calibration package to create this file -->
        <param name="camera_info_url" if="$(arg left_camera_calibrated)" value="file://$(env HOME)/.ros/camera_info/$(arg left_camera_serial).yaml" />
      </node>

    </group>

    <group ns="right">
      <!-- Camera nodelet -->
      <node pkg="nodelet" type="nodelet" name="camera_nodelet" args="load spinnaker_camera_driver/SpinnakerCameraNodelet /camera_nodelet_manager">

        <param name="mode" value="secondary" />
        <param name="frame_id" value="camera_right" />
        <param name="serial" value="$(arg right_camera_serial)" />

        <param name="image_format_color_coding" value="$(arg encoding)" />
        <param name="auto_white_balance" value="$(arg color_balance)" />

        <!-- Use the camera_calibration package to create this file -->
        <param name="camera_info_url" if="$(arg right_camera_calibrated)" value="file://$(env HOME)/.ros/camera_info/$(arg right_camera_serial).yaml" />
      </node>

    </group>

  </group>
</launch>
